### DISCUSSION TITLE
如何让xray使用pac模式？

### DISCUSSION THREAD

**Discussion by sanmaodo** [XTLS/Xray-core#3349] [Q&A] [answered]
刚刚接触xray，想做一个简单Gui客户端，运行在windows下，到pac这个环节卡住了，搜索不到资料，期望有人指点一下，不胜感激。

**Comment by vrnobody** [comment, marked as answer]
PAC是一个js脚本，里面有一个FindProxyForURL()函数。支持PAC的程序，比如浏览器，调用这个函数后会得到代理地址或者`DIRECT`。所以不是xray使用PAC，而是PAC返回xray的代理地址。

很多程序的PAC设置项是一个URL地址，所以需要自行开启一个HTTP服务器。其它程序访问这个URL时，回复的`ContentType`是`application/x-ns-proxy-autoconfig`，内容是PAC的js脚本。注意`ContentType`不要加`charset`，否则部分程序会出问题。

windows的系统代理也可以设置成PAC的URL地址，网上的代码一般通过修改注册表实现。修改系统代理设置这个操作容易引起Microsoft Defender报病毒。系统代理设置对应用商店中的UWP程序无效，好像要解除回环限制还是什么，不太清楚。

**Reply by xiaorong61** [reply]
tun 才是最方便的

**Comment by sanmaodo** [comment]
@vrnobody   感谢！很高兴这么快得到你的答复，为我清晰的指明了方向，虽然暂时我不知道如何开启HTTP服务器，但这应该不会是很难的事情。
>需要自行开启一个HTTP服务器。其它程序访问这个URL时，回复的ContentType是application/x-ns-proxy-autoconfig，内容是PAC的js脚本。

正如你说的，我也是用修改注册表的方式来设置windows系统代理，只需把pac链接写入下面的注册表地址即生效，不需要管理员权限。
>"HKEY_CURRENT_USER\Software\Microsoft\Windows\CurrentVersion\Internet Settings", "AutoConfigURL"`
